src/community/wmts-multi-dimensional/src/main/java/org/geoserver/gwc/wmts/dimensions/HistogramUtils.java

```diff
@@ -0,0 +1,221 @@
+        }
+    }
+
+    /**
+     * Helper method that just founds the histogram type based on domains values.
+     */
+    private static HistogramType findHistogramType(TreeSet<?> domainValues) {
+        Object value = domainValues.first();
+        if (value instanceof Range) {
+            // this is a range so lets use the min value
+            value = ((Range) value).getMinValue();
+        }
+        // let's try to find this histogram type
+        if (value instanceof Number) {
+            return HistogramType.NUMERIC;
+        }
+        if (value instanceof Date) {
+            return HistogramType.TIME;
+        }
+        // well by default we consider the histogram to be of type enumerated
+        return HistogramType.ENUMERATED;
+    }
+
+    /**
+     * Helper method that creates buckets for a numeric domain based on the provided resolution. The returned tuple
+     * will contain the domain representation and the domain buckets.
+     */
+    private static Tuple<String, List<Range>> getNumericBuckets(TreeSet<?> domainValues, String resolution) {
+        Tuple<Double, Double> minMax = DimensionsUtils.getMinMax(domainValues, Double.class);
+        resolution = resolution != null ? resolution : NUMERICAL_DEFAULT_RESOLUTION;
+        double finalResolution = Double.parseDouble(resolution);
+        int i = 0;
+        while ((minMax.second - minMax.first) / finalResolution >= HISTOGRAM_MAX_THRESHOLD && i < MAX_ITERATIONS) {
+            finalResolution += 10;
+            i++;
+        }
+        String domainString = minMax.first.toString() + "/" + minMax.second + "/" + finalResolution;
+        if ((minMax.second - minMax.first) / finalResolution == 1) {
+            return Tuple.tuple(domainString, Collections.singletonList(NumberRange.create(minMax.first, minMax.second)));
+        }
+        List<Range> buckets = new ArrayList<>();
+        for (double step = minMax.first; step < minMax.second; step += finalResolution) {
+            double limit = step + finalResolution;
+            if (limit > minMax.second) {
+                buckets.add(NumberRange.create(step, minMax.second));
+                break;
+            }
+            buckets.add(NumberRange.create(step, limit));
+        }
+        return Tuple.tuple(domainString, buckets);
+    }
+
+    /**
+     * Helper method that creates buckets for a time domain based on the provided resolution. The returned tuple
+     * will contain the domain representation and the domain buckets.
+     */
+    private static Tuple<String, List<Range>> getTimeBuckets(TreeSet<?> domainValues, String resolution) {
+        Tuple<Date, Date> minMax = DimensionsUtils.getMinMax(domainValues, Date.class);
+        resolution = resolution != null ? resolution : TIME_DEFAULT_RESOLUTION;
+        Tuple<String, List<Date>> intervals = getDateIntervals(minMax, resolution);
+        int i = 0;
+        while (intervals.second.size() >= HISTOGRAM_MAX_THRESHOLD && i < MAX_ITERATIONS) {
```

Comment:
There is some logic error here, running a histogram request against time a domain looking like:

````
<Domains xmlns="http://demo.geo-solutions.it/share/wmts-multidim/wmts_multi_dimensional.xsd" xmlns:ows="http://www.opengis.net/ows/1.1">
<DimensionDomain>
<ows:Identifier>elevation</ows:Identifier>
<Domain>
0.0,200.0,400.0,600.0,800.0,1000.0,1200.0,1400.0,1600.0,1800.0,2000.0,3000.0,4000.0,5000.0,6000.0,7000.0,8000.0,9000.0
</Domain>
<Size>18</Size>
</DimensionDomain>
<DimensionDomain>
<ows:Identifier>time</ows:Identifier>
<Domain>
2016-02-23T03:00:00.000Z,2016-02-23T06:00:00.000Z,2016-02-23T09:00:00.000Z,2016-02-23T12:00:00.000Z,2016-02-23T15:00:00.000Z,2016-02-23T18:00:00.000Z,2016-02-23T21:00:00.000Z,2016-02-24T00:00:00.000Z,2016-02-24T03:00:00.000Z,2016-02-24T06:00:00.000Z,2016-02-24T09:00:00.000Z,2016-02-24T12:00:00.000Z,2016-02-24T15:00:00.000Z,2016-02-24T18:00:00.000Z,2016-02-24T21:00:00.000Z,2016-02-25T00:00:00.000Z,2016-02-25T03:00:00.000Z,2016-02-25T06:00:00.000Z,2016-02-25T09:00:00.000Z,2016-02-25T12:00:00.000Z,2016-02-25T15:00:00.000Z,2016-02-25T18:00:00.000Z,2016-02-25T21:00:00.000Z,2016-02-26T00:00:00.000Z,2016-02-26T03:00:00.000Z,2016-02-26T06:00:00.000Z,2016-02-26T09:00:00.000Z,2016-02-26T12:00:00.000Z,2016-02-26T15:00:00.000Z,2016-02-26T18:00:00.000Z,2016-02-26T21:00:00.000Z,2016-02-27T00:00:00.000Z,2016-02-27T03:00:00.000Z,2016-02-27T06:00:00.000Z,2016-02-27T09:00:00.000Z,2016-02-27T12:00:00.000Z,2016-02-27T15:00:00.000Z,2016-02-27T18:00:00.000Z,2016-02-27T21:00:00.000Z,2016-02-28T00:00:00.000Z,2016-02-28T03:00:00.000Z,2016-02-28T06:00:00.000Z,2016-02-28T09:00:00.000Z,2016-02-28T12:00:00.000Z,2016-02-28T15:00:00.000Z,2016-02-28T18:00:00.000Z,2016-02-28T21:00:00.000Z,2016-02-29T00:00:00.000Z
</Domain>
<Size>48</Size>
</DimensionDomain>
<DimensionDomain>
<ows:Identifier>REFERENCE_TIME</ows:Identifier>
<Domain>
2016-02-23T00:00:00.000Z,2016-02-24T00:00:00.000Z,2016-02-25T00:00:00.000Z,2016-02-26T00:00:00.000Z
</Domain>
<Size>4</Size>
</DimensionDomain>
</Domains>
````

Results in:

````
<Histogram xmlns="http://demo.geo-solutions.it/share/wmts-multidim/wmts_multi_dimensional.xsd" xmlns:ows="http://www.opengis.net/ows/1.1">
<ows:Identifier>time</ows:Identifier>
<Domain>
2016-02-23T03:00:00.000Z/2016-02-29T00:00:00.000Z/P1M
</Domain>
<Values>48</Values>
</Histogram>
````

I believe the call to getDateIntervals should have been something else, getDateIntervals always returns a "second" with just one element.